File: resnet/cpp/predict.cpp
#include <sys/stat.h>
#include <iostream>
#include <string>
#include <vector>
#include <torch/torch.h>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>

#include "model.h"

const std::string kSavedModelNamePrefix = "resnet";

using namespace torch::nn;
namespace F = torch::nn::functional;


std::vector<std::string> convert_argv_to_strings(int argc, char** argv)
{
    std::vector<std::string> s;
    for (int i = 0; i < argc; ++i) {
        s.push_back(argv[i]);
    }

    return s;
}

void print_usage_and_exit(int exit_code) {
    std::cout << "Usage: predict" << std::endl;
    std::cout << "    -i <image_to_predict> -m <saved_model_path>" << std::endl;
    std::cout << "    -h" << std::endl;

    exit(exit_code);
}

struct CommandArguments {
    std::string image_to_predict;
    std::string saved_model_path;
};


auto main(int argc, char* argv[]) -> int
{
    // Parse arguments.
    if (argc < 2) {
        print_usage_and_exit(1);
    }

    std::vector<std::string> argv_strings = convert_argv_to_strings(argc, argv);
    CommandArguments args;
    if (argv_strings[1] == "-i") {
        if (argc != 5) {
            print_usage_and_exit(1);
        }
        args.image_to_predict = argv_strings[2];
        if (argv_strings[3] != "-m") {
            print_usage_and_exit(1);
        }
        args.saved_model_path = argv_strings[4];
    } else if (argv_strings[1] == "-h") {
        print_usage_and_exit(0);
    } else {
        print_usage_and_exit(1);
    }

    // Load image.
    cv::Mat mat = cv::imread(args.image_to_predict, cv::IMREAD_GRAYSCALE);
    cv::resize(mat, mat, cv::Size(28, 28));

    // Pre-process.
    cv::Mat mat_float;
    mat.convertTo(mat_float, CV_32F);
    mat_float /= 255.0;

    torch::TensorOptions options(torch::kFloat32);
    torch::Tensor input_tensor = torch::from_blob(
        mat_float.data, {1, 1, 28, 28}, options);

    torch::data::transforms::Normalize<> transforms(0.1307, 0.3081);
    input_tensor = transforms(input_tensor);

    // Create device.
    torch::DeviceType device_type;
    if (torch::cuda::is_available()) {
        std::cout << "Train on GPU." << std::endl;
        device_type = torch::kCUDA;
    } else {
        std::cout << "Train on CPU." << std::endl;
        device_type = torch::kCPU;
    }
    torch::Device device(device_type);

    // Load model.
    std::string model_path = args.saved_model_path + "/"
        + kSavedModelNamePrefix + "_model.pth";
    std::string optimizer_path = args.saved_model_path + "/"
        + kSavedModelNamePrefix + "_optimizer.pth";

    ResNet50 model;
    torch::optim::Adam optimizer(
        model->parameters(), torch::optim::AdamOptions(0.01));

    torch::load(model, model_path);
    torch::load(optimizer, optimizer_path);
    model->to(device);

    // Predict.
    model->eval();
    input_tensor = input_tensor.to(device);
    auto output = model(input_tensor);
    auto pred = output.argmax(1, true);
    std::cout << "Predict: " << pred.cpu()[0][0].template item<int>()
              << std::endl;

  return 0;
}

File: resnet/python/predict.py
import argparse
import torch
import torchvision
import numpy as np
import cv2

from model import ResNet50

SAVED_MODEL_NAME = "resnet.pth"


def main():
    # Parse arguments.
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-i", dest="image_file_to_predict", type=str,
        help="Image file to predict", required=True)
    parser.add_argument(
        "-m", dest="saved_model_path", type=str,
        help="Path to saved model", required=True)
    args = parser.parse_args()

    # Load image.
    img = cv2.imread(args.image_file_to_predict, cv2.IMREAD_GRAYSCALE)

    # Pre-process.
    img = cv2.resize(img, (28, 28))
    transform = torchvision.transforms.Compose([
        torchvision.transforms.ToTensor(),
        torchvision.transforms.Normalize((0.1307,), (0.3081,)),
    ])
    input_tensor = transform(img).unsqueeze(0)

    # Create device.
    if torch.cuda.is_available():
        device_type = "cuda"
        print("Predict on GPU.")
    else:
        device_type = "cpu"
        print("Predict on CPU.")
    device = torch.device(device_type)

    # Load model.
    model_path = "{}/{}".format(args.saved_model_path, SAVED_MODEL_NAME)
    model = ResNet50()
    model.load_state_dict(torch.load(model_path, map_location=device))
    model.to(device)

    # Predict.
    model.eval()
    input_tensor = input_tensor.to(device)
    output = model(input_tensor)
    pred = output.argmax(1, keepdim=True)
    print("Predict: {}".format(pred.cpu()[0][0]))
 

if __name__ == "__main__":
    main()
